book/src/rust/rust_style_guide.md

```diff
@@ -0,0 +1,660 @@
+
+```rust
+fn foo_bar(a: Bar, b: Bar,
+           c: |Bar|) -> Bar {
+    ...
+}
+
+// Same line is fine:
+foo_bar(x, y, |z| { z.transpose(y) });
+
+// Indented body on new line is also fine:
+foo_bar(x, y, |z| {
+    z.quux();
+    z.rotate(x)
+})
+```
+
+##### Alignment
+
+Idiomatic code should not use extra whitespace in the middle of a line to provide alignment.
+
+```rust
+// Good
+struct Foo {
+    short: f64,
+    really_long: f64,
+}
+
+// Bad
+struct Bar {
+    short:       f64,
+    really_long: f64,
+}
+
+// Good
+let a = 0;
+let radius = 7;
+
+// Bad
+let b        = 0;
+let diameter = 7;
+```
+
+#### Braces, semicolons, and commas
+
+Opening braces always go on the same line.
+
+```rust
+fn foo() {
+    ...
+}
+
+fn bar(a: Bar, b: Bar,
+              c: Bar, d: Bar)
+              -> Bar {
+    ...
+}
+
+trait Bar {
+    fn baz(&self);
+}
+
+impl Bar for Baz {
+    fn baz(&self) {
+        ...
+    }
+}
+
+frob(|x| {
+    x.transpose()
+})
+```
+
+`match` arms get braces, except for single-line expressions.
+
+```rust
+match foo {
+    bar => baz,
+    quux => {
+        do_something();
+        do_something_else()
+    }
+}
+```
+
+`return` statements get semicolons.
+
+```rust
+fn foo() {
+    do_something();
+
+    if condition() {
+        return;
+    }
+
+    do_something_else();
+}
+```
+
+##### Trailing commas
+
+> Opossible rule: a trailing comma should be included whenever the closing delimiter appears on a separate line:
+
+```rust
+Foo { bar: 0, baz: 1 }
+
+Foo {
+    bar: 0,
+    baz: 1,
+}
+
+match a_thing {
+    None => 0,
+    Some(x) => 1,
+}
+```
+
+
+#### The Cargo.toml File
+
+The Cargo.toml file should adhere to the following template:
+
+```toml
+[package]
+name = "..."
+version = "..."
+edition = "..."
+authors = ["IOG <contact@iohk.io>"]
+readme = "README.md"
+license-file = "LICENSE"
+repository = "https://github.com/input-output-hk/repo-name"
+homepage = "https://github.com/input-output-hk/repo-name"
+documentation = "https://github.com/input-output-hk/repo-name"
+categories = ["..."]
+keywords = ["..."]
+description = "..."
+publish = false
+
+[package.metadata.docs.rs]
+# To build locally:
+# RUSTDOCFLAGS="--cfg doc_cfg" cargo +nightly doc --all-features --open
+all-features = true
+rustdoc-args = ["--cfg", "doc_cfg"]
+
+[badges]
+is-it-maintained-issue-resolution = { repository = "input-output-hk/repo-rs" }
+is-it-maintained-open-issues = { repository = "input-output-hk/repo-rs" }
+maintenance = { status = "actively-developed" }
+
+[[bin]]
+...
+
+[features]
+...
+
+[dependencies]
+...
+
+[dev-dependencies]
+...
+
+[build-dependencies]
+...
+
+[profile....]
+...
+```
+Specifically, we have:
+
+1. Use double quotes instead of single quotes.
+2. Use the above as the standard ordering of the [package] map.
+3. [[bin]] before [features] before [dependencies] before [dev-dependencies] before [build-dependencies] before [profile] settings.
+4. Order features and dependencies alphabetically.
```

Comment:
I kind of agree,  but i would go further.  I think dependencies should have a comment stating WHY its being used.  My experience is as projects grow they gather dependencies like flies to a rotting carcass.  And no one is brave enough to remove them or look at them in detail because they don;t know why they are a dependency.  This also goes for features,  why do we use/need those features.

So,  building on what @cameron1024 said,  I would prefer:

```toml
[dependencies] 

# Local Crates
chain-addr = { path = "../../chain-libs/chain-addr" }  # Vote storage chain address type definition

# Tokio Async-IO Framework
#     Rest API endpoints are async.
tokio = { version = "^1.15", features = ["rt-multi-thread", "time", "sync", "rt", "signal", "test-util"] } # Why we use these features.
tokio-stream = { version = "0.1.4", features = ["sync"] }  # Feature "sync" is used because ...
tokio-util = { version = "0.6.0", features = ["time"] } # Feature "time" is used because ... 
```

So groups of related dependencies are clear and also why the choice what the program is using the crate for is clear.

Reply:
Possibly not for this document, but we should have a list of crates we use by preference, so that when someone needs a crate they look at the curated list and they use the one in that, that provides what they need.  Then, if they add a new crate, they should update that list with the crate, and what they use it for.  If they need to add a new crate, it should entail a brief discussion on slack, especially when there are options, before diving in.  

I find in rust there are usually 2 or more crates to do the same thing, of varying quality.  Its really easy to end up using different crates for the same things all over the place in a big code base.

Reply:
And in review,  this should be checked.  
1. Did a new crate get added as a dependency thats not in the list of "used crates", if so why.
2. Is someone re-inventing the wheel (ie, writing code for something when a perfectly usable crate already exists which does it.)

Reply:
I think we can add a new section or a new check item into the `pull_request_template.md`, where it is important to add which new dependencies are coming with this new PR.

Reply:
Created https://github.com/input-output-hk/catalyst-core/pull/207